### ISSUE TITLE
Content-Length isn't recorded or returned

### ISSUE THREAD

**Issue by Fishbowler** [SpectoLabs/hoverfly#697]
My organisation has a disperate collection of APIs consumed in combinations from various front-ends. Hoverfly could simplify testing by simulating one or more of these to focus on the "thing under test".

I've already got a set of API tests in place, and I'm using them to power the recording and test the simulation.

One endpoint accepts and validates some post-data, and when the data is correct, returns a 200 and no body. The API test validates that the Content-Length header is 0.

When recording through Hoverfly using the `--all-headers` switch, the Content-Length header isn't captured, and when simulating, the test fails. I've manually edited the export to include the header, but I'm not seing any change.

```
				"response": {
					"status": 200,
					"body": "",
					"encodedBody": true,
					"headers": {
						"Cache-Control": [
							"private"
						],
						"Content-Encoding": [
							"gzip"
						],
						"Date": [
							"Tue, 13 Feb 2018 08:25:14 GMT"
						],
						"Hoverfly": [
							"Was-Here"
						],
						"Last-Modified": [
							"Tue, 13 Feb 2018 08:25:14 GMT"
						],
						"Link": [
							""
						],
						"Transfer-Encoding": [
							"chunked"
						],
						"Vary": [
							"Accept-Encoding"
						],
						"Content-Length": [
							"0"
						]
					},
					"templated": false
				}
```

Is this header specifically excluded? Or excluded when 0?
Any advice on how to make this more "faithful"?

I'm using hoverctl and hoverfly v0.15.0, downloaded binaries, on the commandline.

**Comment by tommysitu** [comment]
Interesting problem. Hoverfly is default to not capture any Request headers, but it should capture all Response headers. The '--all-headers' flag is to tell hoverfly to capture all Request headers therefore it won't make any difference here. 

Could you confirm that the response indeed contains the Content-Length header? For example making a request with curl and -v flag without hoverfly as a proxy.

**Comment by Fishbowler** [comment]
I'm in a particularly useful scenario in that I'm fueling my recordings through a set of API tests.
With hoverfly, the Content-Length is missing and the test fails.
Without hoverfly, the Content-Lenght is present and 0 and the test passes.

**Comment by tommysitu** [comment]
Ok, in that case I will do some debuggings when I have my laptop around next week.

**Comment by tommysitu** [comment]
Hi @Fishbowler, I just realized this issue is possibly a duplicate of this one: https://github.com/SpectoLabs/hoverfly/issues/422

The `Transfer-Encoding`  and `Content-Length` headers are mutually exclusive apparently, please see  https://stackoverflow.com/questions/3304126/chunked-encoding-and-content-length-header

Hoverfly doesn't alter any response headers which I just tested, so it must be caused by something else.

**Comment by tommysitu** [comment]
Hi @Fishbowler, I can reproduce the problem now. If I made a request with curl: 
```
curl -X POST 'https://api.twilio.com/2010-04-01/Accounts/AC3d0e9aaf91e037d5f0038248b1234b55/Messages.json' --data 'To=%2B15558675309&From=%2B15017250604&Body=This%20is%20the%20ship%20that%20made%20the%20Kessel%20Run%20in%20fourteen%20parsecs%3F&MediaUrl=https://c1.staticflickr.com/3/2899/14341091933_1e92e62d12_b.jpg' -v
```
The response contains `Content-Length` header: 
```
< HTTP/1.1 401 UNAUTHORIZED
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Headers: Accept, Authorization, Content-Type, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since
< Access-Control-Allow-Methods: GET, POST, DELETE, OPTIONS
< Access-Control-Allow-Origin: *
< Access-Control-Expose-Headers: ETag
< Content-Type: application/json
< Date: Tue, 27 Feb 2018 10:28:49 GMT
< Twilio-Request-Duration: 0.003
< Twilio-Request-Id: RQd30ca477434341a2a8baac8f5f1f6f22
< WWW-Authenticate: Basic realm="Twilio API"
< X-Powered-By: AT-5000
< X-Shenanigans: none
< Content-Length: 205
< Connection: keep-alive
```
But with Hoverfly as proxy, the `Transfer-Encoding` header is added to the response instead: 
```
* upload completely sent off: 202 out of 202 bytes
< HTTP/1.1 401 UNAUTHORIZED
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Headers: Accept, Authorization, Content-Type, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since
< Access-Control-Allow-Methods: GET, POST, DELETE, OPTIONS
< Access-Control-Allow-Origin: *
< Access-Control-Expose-Headers: ETag
< Connection: keep-alive
< Content-Type: application/json
< Date: Tue, 27 Feb 2018 10:34:08 GMT
< Hoverfly: Was-Here
< Transfer-Encoding: chunked
< Twilio-Request-Duration: 0.002
< Twilio-Request-Id: RQ308d5aa2452248e5adbc8276b86f3868
< Www-Authenticate: Basic realm="Twilio API"
< X-Powered-By: AT-5000
< X-Shenanigans: none
```

**Comment by luc2016** [comment]
For me, it's the request that has been altered by hoverfly proxy.
The original request contains `"Content-Length": ["0"]`, when go though hoverfly proxy, the proxy removed the content-length header but added `Transfer-Encoding: chunked` instead. Which cause the server failed with `411` error.

**Comment by benjih** [comment]
Hey @luc2016,

Been looking at this issue trying to get to the bottom of it. I was able to recreate this bug and it appears to only be happening with responses being served over HTTPS. Is this the same for you?

I have pushed a branch called [goproxy-fix](/SpectoLabs/hoverfly/tree/goproxy-fix) that should fix this issue. Would it be possible for you to build this branch and let me know if it fixes it for you, @luc2016?
